src/client/QXmppTrustMemoryStorage.cpp

```diff
@@ -0,0 +1,322 @@
+
+        for (const auto &key : keys) {
+            d->processedKeys.remove(encryption, key);
+        }
+    }
+}
+
+QFuture<QList<QString>> QXmppTrustMemoryStorage::keys(const QString &encryption) const
+{
+    QList<QString> keys;
+
+    for (const auto &key : d->processedKeys.values(encryption)) {
+        keys.append(key.id);
+    }
+
+    return QXmpp::Private::makeReadyFuture(std::move(keys));
+}
+
+QFuture<QList<QString>> QXmppTrustMemoryStorage::keys(const QString &encryption, const QString &keyOwnerJid) const
+{
+    QList<QString> keys;
+
+    for (const auto &key : d->processedKeys.values(encryption)) {
+        if (key.ownerJid == keyOwnerJid) {
+            keys.append(key.id);
+        }
+    }
+
+    return QXmpp::Private::makeReadyFuture(std::move(keys));
+}
+
+QFuture<QList<QString>> QXmppTrustMemoryStorage::keys(const QString &encryption, const QString &keyOwnerJid, const TrustLevel trustLevel) const
+{
+    QList<QString> keys;
+
+    for (const auto &key : d->processedKeys.values(encryption)) {
+        if (key.ownerJid == keyOwnerJid && key.trustLevel == trustLevel) {
+            keys.append(key.id);
+        }
+    }
+
+    return QXmpp::Private::makeReadyFuture(std::move(keys));
+}
+
+QFuture<QMultiHash<QString, QString>> QXmppTrustMemoryStorage::keys(const QString &encryption, const QXmppTrustStorage::TrustLevel trustLevel) const
+{
+    QMultiHash<QString, QString> keys;
+
+    for (const auto &key : d->processedKeys.values(encryption)) {
+        if (key.trustLevel == trustLevel) {
+            keys.insert(key.ownerJid, key.id);
+        }
+    }
+
+    return QXmpp::Private::makeReadyFuture(std::move(keys));
+}
+
+QFuture<QList<QString>> QXmppTrustMemoryStorage::trustedKeys(const QString &encryption, const QString &keyOwnerJid) const
+{
+    QList<QString> keys;
+
+    for (const auto &key : d->processedKeys.values(encryption)) {
+        if (key.ownerJid == keyOwnerJid &&
+            (key.trustLevel == TrustLevel::AutomaticallyTrusted ||
+             key.trustLevel == TrustLevel::ManuallyTrusted ||
+             key.trustLevel == TrustLevel::Authenticated)) {
+            keys.append(key.id);
+        }
+    }
+
+    return QXmpp::Private::makeReadyFuture(std::move(keys));
+}
+
+QFuture<QList<QString>> QXmppTrustMemoryStorage::keyOwnersWithAuthenticatedKeys(const QString &encryption) const
+{
+    QList<QString> keyOwners;
+
+    for (const auto &key : d->processedKeys.values(encryption)) {
+        if (key.trustLevel == TrustLevel::Authenticated) {
+            keyOwners.append(key.ownerJid);
+        }
+    }
+
+    return QXmpp::Private::makeReadyFuture(std::move(keyOwners));
+}
+
+void QXmppTrustMemoryStorage::setTrustLevel(const QString &encryption, const QString &keyOwnerJid, const QList<QString> &keyIds, const TrustLevel trustLevel)
+{
+    for (auto keyId : keyIds) {
+        auto isKeyFound = false;
+
+        for (auto i = d->processedKeys.find(encryption); i != d->processedKeys.end(); ++i) {
```

Comment:
```suggestion
        for (auto itr = d->processedKeys.find(encryption); itr != d->processedKeys.end(); ++itr) {
```
otherwise looks like an index